src/swarm/neo/request/RequestEventDispatcher.d

```diff
@@ -0,0 +1,767 @@
+                events over
+
+    ***************************************************************************/
+
+    private void handleNextEvent ( RequestOnConnBase.EventDispatcher conn )
+    {
+        Const!(void)[] received;
+
+        auto resume_code = conn.receiveAndHandleEvents(
+            ( in void[] recv_payload )
+            {
+                received = recv_payload;
+            });
+
+        // Interesting to note that resume_code:0 has no meaning here
+        // and could be used by the request implementation, in theory.
+        this.dispatchSignal(conn, resume_code);
+
+        if ( received.length )
+            this.dispatchReceivedPayload(conn, received);
+    }
+
+    /***************************************************************************
+
+        Registers `fiber` as awaiting `event`.
+
+        Params:
+            fiber = fiber to register as awaiting an event
+            event = event which fiber is awaiting
+
+        Throws:
+            1. if a message is being awaited and that message type is already
+               registered as being awaited (by the same or another fiber) --
+               each message type may only be handled by one fiber.
+            2. if a signal is being awaited and that code is already registered
+               as being awaited (by the same or another fiber) -- each signal
+               may only be handled by one fiber.
+            3. if sending a payload is being awaited and the fiber is already
+               awaiting another payload being sent -- each fiber may only send
+               one payload at a time.
+
+    ***************************************************************************/
+
+    private void register ( MessageFiber fiber, EventIn event )
+    {
+        with ( event.Active ) switch ( event.active )
+        {
+            case message:
+                foreach ( waiting_fiber; this.waiting_fibers )
+                    if ( waiting_fiber.event.active == message )
+                        enforce(waiting_fiber.event.message.type
+                            != event.message.type,
+                            "Only one fiber may handle each message type");
+                break;
+            case signal:
+                foreach ( waiting_fiber; this.waiting_fibers )
+                    if ( waiting_fiber.event.active == signal )
+                        enforce(waiting_fiber.event.signal.code
+                            != event.signal.code,
+                            "Only one fiber may handle each signal");
+                break;
+            case send:
+                foreach ( waiting_fiber; this.waiting_fibers )
+                    if ( waiting_fiber.event.active == send )
+                        enforce(waiting_fiber.fiber != fiber,
+                            "Each fiber may only send one thing at a time");
+                break;
+            default:
+                assert(false);
+        }
+
```

Comment:
Interesting case, comparing two unions: Looks terribly redundant, but I don't see how it could be simplified.